File: Application/src/main/java/com/shareyourproxy/widget/CustomEditText.java
package com.shareyourproxy.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputConnection;
import android.view.inputmethod.InputConnectionWrapper;
import android.widget.EditText;

import com.shareyourproxy.api.rx.RxBusDriver;
import com.shareyourproxy.api.rx.event.TextViewEditorActionEvent;

/**
 * Created by Evan on 11/2/15.
 */
public class CustomEditText extends EditText {
    private RxBusDriver _rxBus;

    public CustomEditText(Context context) {
        super(context);
    }

    public CustomEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CustomEditText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public InputConnection onCreateInputConnection(EditorInfo outAttrs) {
        return new CustomInputConnection(_rxBus, super.onCreateInputConnection(outAttrs), true);
    }

    public void setRxBus(RxBusDriver rxBus) {
        _rxBus = rxBus;
    }

    private static class CustomInputConnection extends InputConnectionWrapper {
        private final RxBusDriver _rxBus;

        public CustomInputConnection(RxBusDriver rxBus, InputConnection target, boolean mutable) {
            super(target, mutable);
            _rxBus = rxBus;
        }

        @Override
        public boolean deleteSurroundingText(int beforeLength, int afterLength) {
            if (afterLength == 0) {
                _rxBus.post(new TextViewEditorActionEvent(
                    new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DEL)));
            }
            return super.deleteSurroundingText(beforeLength, afterLength);
        }
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/widget/CustomEditText.kt
package com.shareyourproxy.widget

import android.content.Context
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.KeyEvent.ACTION_DOWN
import android.view.KeyEvent.KEYCODE_DEL
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputConnection
import android.view.inputmethod.InputConnectionWrapper
import android.widget.EditText
import com.shareyourproxy.api.rx.RxBusRelay.post
import com.shareyourproxy.api.rx.event.TextViewEditorActionEvent

/**
 * Custom Edit Text for listening to delete key actions.
 */
class CustomEditText : EditText {

    constructor(context: Context) : super(context) {
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
    }

    override fun onCreateInputConnection(outAttrs: EditorInfo): InputConnection {
        return CustomInputConnection(super.onCreateInputConnection(outAttrs), true)
    }

    private class CustomInputConnection(target: InputConnection, mutable: Boolean) : InputConnectionWrapper(target, mutable) {
        override fun deleteSurroundingText(beforeLength: Int, afterLength: Int): Boolean {
            if (afterLength == 0) {
                post(TextViewEditorActionEvent(KeyEvent(ACTION_DOWN, KEYCODE_DEL)))
            }
            return super.deleteSurroundingText(beforeLength, afterLength)
        }
    }
}
